### DISCUSSION TITLE
NGINX configurazione per porta 443

### DISCUSSION THREAD

**Discussion by Matty1184** [indomus/forum#2244] [Networking]
Ciao a tutti, ho un dubbio sulla configurazione che ho fatto con il proxy NGINX.
attualmente raggiungo HA dall'esterno con server proxy NGINX e DUckDNS su https:\\miodominio.duckdns:miaporta. facciamo esempio che la porta sia per convenzione la numero 2023 per fare un esempio concreto quindi https:\\miodominio.duckdns:2023
sul router ho aperto le porte 80, 443, 2023
forwarding 80 su 80, 443 su 443 e 2023 su 8123 di HA.
nella configurazione di NGINX nella sezione "Modifica le porte sul tuo host che sono esposte dal componente aggiuntivo" ho inserito 2023. 
raggiungo da esterno con ssl tutto ok.
Stavo guardando di integrare Alexa e leggevo che devo avere attivo il forwarding della porta 8123 in locale sulla porta 443 esterna di Home Assistant. 
premesso che non me intendo tantissimo di reti, se cosi fosse, come richiesto dovrei fare il forwarding della 443 su 8123 e io ora ho la configurazione sbagliata o no? io interpreto che dovrei avere il forwarding impostato su 80 su 80, 443 su 8123 di HA e raggiungere di conseguenza dall'esterno https:\\miodominio.duckdns senza porta oppure 443.
oppure basta modificare la porta i NGINX nella sezione "Modifica le porte sul tuo host che sono esposte dal componente aggiuntivo" mettendo la 443? (se non ho capito male NGINX non reindirizza in automatico alla porta di HA) 
corretto la modifica che dovrei fare o mi sto perdendo?

**Comment by snakuzzo** [comment]
Dipende dal tipo di integrazione che usi per interfacciare Amazon (Alexa) con la tua istanza di HA.
Se usi haaska non è necessario esporre HA sulla 443... per altro tipo di integrazione ricordo che invece era necessario esporre HA sulla 443.

**Reply by Matty1184** [reply]
Ok grazie per l'info, sicuramente allora opterò per quella soluzione.
il mio ragionamento sulle porte però è corretto? idem sul principio di NGINX?

**Reply by snakuzzo** [reply]
Premesso che, nel caso si utilizzi un reverse proxy, i certificati non vanno specificati nel `configuration.yaml` di HA...HA dovrà restare configurato in HTTP senza i certificati.
Detto ciò, ammesso che tu abbia fatto tutto "come cristo comanda", le regole di port forwarding che hai specificato sono errate...il fatto che tu dica "raggiungo HA dall'esterno con server proxy NGINX e DUckDNS" può essere vero, ma non vuol dire che tu lo faccia in maniera corretta...e dico questo proprio perché una delle regole che hai indicato non mi torna affatto.

Partiamo dall'addon. Ci sono due addon relativi ad NGINX, immagino tu stia utilizzando "NGINX Home Assistant SSL proxy".
L'addon non si preoccupa di fare le challenge per il rinnovo automatico dei certificati, ne consegue che nell'addon, per quanto riguarda le porte, hai necessità solo di configurare la porta da utilizzare per esporre HA.

Ammettiamo che HA di suo risponda in locale in HTTP sulla porta 8123 (che è la configurazione di default)

NGINX, che fondamentalmente è un altro container docker, risponderà sulla porta che tu gli specifichi nella configurazione. Ammettiamo che io esponga la 5223

![image](https://user-images.githubusercontent.com/7205371/210814429-99a0b30e-ac2f-4a40-b88e-df49f1f29f40.png)

Se da terminale lanci il comando...

```
docker ps -a | grep nginx
```

...puoi vedere il bind delle porte dei vari container docker...e come vedi, tra quelle di NGINX c'è la 5223 (nel tuo caso sarà 2023)...

![image](https://user-images.githubusercontent.com/7205371/210814117-065224e6-4fdd-42c3-9a11-581ca50af837.png)

Attenzione, per ora stiamo ragionando sempre in locale...non siamo arrivati ancora al port forwarding.

A questo punto, puntando l'host sulla 5223, raggiungerai l'istanza di HA (che di base è sulla 8123). Attenzione! Il proxy NGINX accetterà chiamate via browser solo se arriveranno con il FQDN specificato quindi, se proprio vuoi toglierti il dubbio e fare una prova, ti basterà forzare la risoluzione nel file hosts della macchina client da cui farai la prova. 

Finita la prova, assicurati di rimuovere la entry dal file hosts, e passiamo al port forwarding.

L'unica cosa che è necessario fare, è creare una regola forwardando il traffico dalla 5223 TCP esterna -> 5223 TCP interna (NGINX)

Le porte 80 e 443 non servono sicuramente ad NGINX, ma potrebbero servire a qualche addon che fa le challenge DNS. L'addon DuckDNS fala DNS challenge e non ha bisogno di forwarding di porte.

Tornando alla regola che ti dicevo che non mi tornava... 2023 -> 8123. 
Questa regola non ti fa atterrare sul proxy NGINX, ma direttamente su HA. Se dici che funziona, vuol dire che hai specificato i certificati su Core...e se vuoi usare NGINX, questa cosa è sbagliata.

Nel tuo caso le regole di port forwarding necessarie saranno:

2023 -> 2023  - che utilizzerà l'addon NGINX

**Reply by Matty1184** [reply]
Perdonami ho scritto una cosa errata sul forwarding del mio router.
io ho forwarding 80 su 80, 443 su 443 e 2023 su 2023
ho una porta di riserva provvisoria che è ipesterno:5000 su 8123 ma la uso solo in casi estremi ora che sto facendo prove per questo avevo scritto quello.
confermo che non ho esposto in HA i certificati ma li gestisce NGINX quindi in sostanza ho fatto come hai descritto, anche la prova del bind porte è come da tua immagine.
la mia domanda da ignorante è:
dal momento che vorrei esporre la 443 in quanto ho alcune app che non dialogano, avere il forwarding 443 su 443 è già attivo e aperto o devo inserire in NGINX la porta 443 anzichè ma mia 2023?

**Reply by snakuzzo** [reply]
ok.
Ma 443 su 443 che cacchio c'entra?
Mica HA o il reverse proxy risponde sulla 443 ?
Per rendere disponibile HA "ANCHE" sulla 443 la regola di port forwarding, anche in base a quello che ti ho spiegato prima, dovrebbe essere...

443 -> 2023 (NGINX)     ... non trovi ?
